src/log/Log.cc

```diff
@@ -40,66 +38,28 @@ static void log_on_exit(void *p)
 }
 
 Log::Log(const SubsystemMap *s)
-  : m_indirect_this(NULL),
+  : m_indirect_this(nullptr),
     m_subs(s),
-    m_queue_mutex_holder(0),
-    m_flush_mutex_holder(0),
-    m_new(), m_recent(),
-    m_fd(-1),
-    m_uid(0),
-    m_gid(0),
-    m_fd_last_error(0),
-    m_syslog_log(-2), m_syslog_crash(-2),
-    m_stderr_log(1), m_stderr_crash(-1),
-    m_graylog_log(-3), m_graylog_crash(-3),
-    m_log_buf(nullptr), m_log_buf_pos(0),
-    m_stop(false),
-    m_max_new(DEFAULT_MAX_NEW),
-    m_max_recent(DEFAULT_MAX_RECENT),
-    m_inject_segv(false)
+    m_recent(DEFAULT_MAX_RECENT)
 {
-  int ret;
-
-  ret = pthread_mutex_init(&m_flush_mutex, NULL);
-  ceph_assert(ret == 0);
-
-  ret = pthread_mutex_init(&m_queue_mutex, NULL);
-  ceph_assert(ret == 0);
-
-  ret = pthread_cond_init(&m_cond_loggers, NULL);
-  ceph_assert(ret == 0);
-
-  ret = pthread_cond_init(&m_cond_flusher, NULL);
-  ceph_assert(ret == 0);
-
-  m_log_buf = (char*)malloc(MAX_LOG_BUF);
-
-  // kludge for prealloc testing
-  if (false)
-    for (int i=0; i < PREALLOC; i++)
-      m_recent.enqueue(new Entry);
+  m_log_buf.reserve(MAX_LOG_BUF);
 }
 
 Log::~Log()
 {
   if (m_indirect_this) {
-    *m_indirect_this = NULL;
+    *m_indirect_this = nullptr;
   }
 
   ceph_assert(!is_started());
   if (m_fd >= 0)
     VOID_TEMP_FAILURE_RETRY(::close(m_fd));
-  free(m_log_buf);
-
-  pthread_mutex_destroy(&m_queue_mutex);
-  pthread_mutex_destroy(&m_flush_mutex);
-  pthread_cond_destroy(&m_cond_loggers);
-  pthread_cond_destroy(&m_cond_flusher);
 }
 
 
 ///
 void Log::set_coarse_timestamps(bool coarse) {
+  Locker lock(m_flush_mutex);
```

Comment:
You mean at this location or everywhere? We are using the unlock/lock methods of Locker in other locations. I don't really see a need to optimize for these methods which are called only once at startup?

Reply:
Also, the condition variables only take a `unique_lock`.

Reply:
> You mean at this location or everywhere?

i mean the wherever `std::scoped_lock` applies.

> We are using the unlock/lock methods of Locker in other locations. 

yes, i knew.

> Also, the condition variables only take a unique_lock.

yeah, `condition_variables` takes `unique_lock`.  in addition to an optimisation, i'd take this as an improvement in readability. when one reads a function accepts `std::string` instead of `const std::string&` or `std::string_view`, he/she would wonder why. the same applies to `unique_lock`. when i am reading the code, i was asking myself, why we wanted to use `unique_lock` here?